src/libraries/System.Runtime.InteropServices/tests/TestAssets/SharedTypes/ComInterfaces/IStatefulCollectionPinnableReference.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.InteropServices;
+using System.Runtime.InteropServices.Marshalling;
+
+namespace SharedTypes.ComInterfaces
+{
+    [GeneratedComInterface(), Guid("0A52BA7C-E08B-42A4-A1F4-1A2BF2C07E60")]
+    internal partial interface IStatefulCollectionPinnableReference
+    {
+        void Method(
+            [MarshalUsing(CountElementName = nameof(size))] StatefulCollectionPinnableReference<int> p,
+            int size);
+
+        void MethodIn(
+            [MarshalUsing(CountElementName = nameof(size))] in StatefulCollectionPinnableReference<int> pIn,
+            in int size);
+
+        void MethodRef(
+            [MarshalUsing(CountElementName = nameof(size))] ref StatefulCollectionPinnableReference<int> pRef,
+            int size);
+
+        void MethodOut(
+            [MarshalUsing(CountElementName = nameof(size))] out StatefulCollectionPinnableReference<int> pOut,
+            out int size);
+
+        [return: MarshalUsing(CountElementName = nameof(size))]
+        StatefulCollectionPinnableReference<int> Return(int size);
+
+        [PreserveSig]
+        [return: MarshalUsing(CountElementName = nameof(size))]
+        StatefulCollectionPinnableReference<int> ReturnPreserveSig(int size);
```

Comment:
I'm fine having this if it works without a lot of ceremony. Our first version of the COM source generator should be strict matching of COM, so `HRESULT` returns. This isn't valid COM unfortunately so if we hit any issues with this scenario it is fine to drop it.